File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    today = [int(i) for i in today.split(".")]
    terms_dict = {}
    answer = []
    cnt = 1
    
    for term in terms:
        term = term.split(" ")
        terms_dict[term[0]] = int(term[-1])
    
    for privacy in privacies:
        privacy = privacy.split(" ")
        date = list(map(int, privacy[0].split(".")))
        year, month, day = date
        month += terms_dict[privacy[1]]
        day -= 1
        
        while month > 12:
            month -= 12
            year += 1
        
        if day == 0:
            month -= 1
            day = 28
            if month == 0:
                year -= 1
                month = 12
        
        print(year, month, day)
        if year - today[0] < 0:
            answer.append(cnt)
        elif year - today[0] == 0:
            if month - today[1] < 0:
                answer.append(cnt)
            elif month - today[1] == 0:
                if day - today[2] < 0:
                    answer.append(cnt)
        
        
        cnt += 1
        
        
    return answer

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
function solution(today, terms, privacies) {
    var answer = [];
    const now = today.split(".").map((el) => Number(el))
    const nowDays = now[0] * 12 * 28 + now[1] * 28 + now[2]
    terms = terms.reduce((acc, cur) => {
       const [type, date] = cur.split(" ")
       return {...acc, [type]: Number(date)}
    } , {})
    privacies.forEach((el, idx) => {
        const [date, type] = el.split(" ")
        let [year, month, day] = date.split(".").map((it) => Number(it))
        let expiration = year * 12 * 28 + month * 28 + day + terms[type] * 28
        if (nowDays >= expiration) answer.push(idx + 1)
    })
    return answer;
}